refactor: Improve exception handling in Adapter components

shyim opened the pull request:
### 1. Why is this change necessary?

This commit introduces a more consistent and structured approach to exception handling in Shopware's Adapter components:

- Added new exception methods in AdapterException for various scenarios
- Replaced direct exception throwing with domain-specific exceptions
- Improved type safety and error reporting in Twig extensions and filesystem adapters
- Standardized error code and message generation for different error types


### 2. What does this change do, exactly?


### 3. Describe each step to reproduce the issue or behaviour.


### 4. Please link to the relevant issues (if any).
<!-- Examples:
- closes #123  - closes the issue #123 when the PR is merged
- relates #123 - relates to the issue #123

In case of issue existing only on Jira, link to the Jira issue.
- Jira issue: https://shopware.atlassian.net/browse/NEXT-123
-->

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfill them.

shyim committed:
refactor: Improve exception handling in Adapter components

This commit introduces a more consistent and structured approach to exception handling in Shopware's Adapter components:

- Added new exception methods in AdapterException for various scenarios
- Replaced direct exception throwing with domain-specific exceptions
- Improved type safety and error reporting in Twig extensions and filesystem adapters
- Standardized error code and message generation for different error types

shyim committed:
feat: Add currency filter error handling in Twig extension

Improve error handling for currency filter by:
- Adding specific exception methods in AdapterException for missing context and ISO code
- Replacing generic InvalidArgumentException with domain-specific AdapterExceptions
- Enhancing error reporting for currency filter processing

shyim committed:
refactor: Make MemoryAdapterFactory final

Improve class design by marking MemoryAdapterFactory as final, preventing further inheritance and ensuring a more predictable implementation.

shyim committed:
test: Improve PrefixFilesystem error handling test

Update PrefixFilesystemTest to use domain-specific AdapterException for empty prefix validation, aligning with recent improvements in exception handling for Adapter components.

shyim committed:
refactor: Cleanup and improve Adapter component code

Perform minor code improvements across Adapter components:
- Use global sprintf function in AdapterException
- Remove unnecessary parameter in CacheValueCompressor
- Remove unused import in FilesystemFactory
- Reorganize imports and improve method signature in CurrencyFilter
- Remove unused import in PrefixFilesystemTest

shyim committed:
refactor: Improve type handling in InstanceOfExtension Twig filter

Modify the isInstanceOf method to accept more flexible input types while maintaining core functionality. This change allows for more lenient type checking in Twig templates while preserving the original method's intent.

mitelg commented on src/Core/Framework/Adapter/AdapterException.php:
```diff
@@ -146,4 +155,81 @@ public static function invalidArgument(string $message): self
+
+    public static function filesystemFactoryNotFound(string $type): AdapterFactoryNotFoundException
+    {
+        return new AdapterFactoryNotFoundException($type);
```
the exception seems not be caught, so it can be removed. please deprecated it and only return an instance of this domain exception

mitelg commented on src/Core/Framework/Adapter/AdapterException.php:
```diff
@@ -146,4 +155,81 @@ public static function invalidArgument(string $message): self
+
+    public static function duplicateFilesystemFactory(string $type): DuplicateFilesystemFactoryException
+    {
+        return new DuplicateFilesystemFactoryException($type);
```
the exception seems not be caught, so it can be removed. please deprecated it and only return an instance of this domain exception

mitelg commented on src/Core/Framework/Adapter/Filesystem/PrefixFilesystem.php:
```diff
@@ -22,7 +23,7 @@ public function __construct(
         string $prefix
     ) {
         if (empty($prefix)) {
-            throw new \InvalidArgumentException('The prefix must not be empty.');
```
if someone relies on catching an InvalidArgumentException, this is a break. please use a feature flag here

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/InstanceOfExtension.php:
```diff
@@ -16,6 +16,10 @@ public function getTests(): array
+     * @param object $var
+     * @param class-string $class
+     */
     public function isInstanceOf($var, $class): bool
```
why is that class not internal :see_no_evil: can we please use native types with the next major or now, if it is non breaking?

mitelg commented on src/Core/Framework/Adapter/Twig/Filter/CurrencyFilter.php:
```diff
@@ -31,7 +32,14 @@ public function getFilters()
     }
 
     /**
+     * @param array<string, mixed> $twigContext
```
also please use native types or "deprecate" them, so we have them in the next major

shyim committed:
Merge remote-tracking branch 'origin/trunk' into fix-phpstan-adapter

shyim committed:
feat(framework): Prepare filesystem and adapter exceptions for v6.8.0

- Modify AdapterException to support new error handling for filesystem factories
- Add deprecation notices for filesystem and adapter-related methods
- Update PrefixFilesystem and test cases to align with new exception handling
- Add deprecation notices for Twig extensions and filters

mitelg commented:
@shyim could you please resolve the conflicts first? then I will have another look :slightly_smiling_face: :v:

shyim committed:
Merge remote-tracking branch 'origin/trunk' into fix-phpstan-adapter

shyim committed:
chore: update baseline

shyim commented:
@mitelg done :)

mitelg commented on src/Core/Framework/Adapter/AdapterException.php:
```diff
@@ -166,4 +177,105 @@ public static function invalidArgument(string $message): self
+     */
+    public static function filesystemFactoryNotFound(string $type): AdapterFactoryNotFoundException|self
+    {
+        if (Feature::isActive('v6.8.0.0')) {
```
can we make it the other way around? so with negotiation, in the major preparation we can then just throw away the whole if block. with this, we would need to think about, what needs to stay and what can be removed :see_no_evil:

mitelg commented on src/Core/Framework/Adapter/AdapterException.php:
```diff
@@ -166,4 +177,105 @@ public static function invalidArgument(string $message): self
+     */
+    public static function duplicateFilesystemFactory(string $type): DuplicateFilesystemFactoryException|self
+    {
+        if (Feature::isActive('v6.8.0.0')) {
```
see above

mitelg commented on src/Core/Framework/Adapter/Filesystem/PrefixFilesystem.php:
```diff
@@ -22,6 +24,11 @@ public function __construct(
         string $prefix
     ) {
         if (empty($prefix)) {
+            if (Feature::isActive('v6.8.0.0')) {
```
see above

shyim committed:
refactor: Adjust exception handling based on feature activation for v6.8.0

mitelg commented on src/Core/Framework/Adapter/AdapterException.php:
```diff
@@ -166,4 +177,105 @@ public static function invalidArgument(string $message): self
+     */
+    public static function filesystemFactoryNotFound(string $type): AdapterFactoryNotFoundException|self
+    {
+        if ((bool) Feature::isActive('v6.8.0.0')) {
```
what is this shit? teach your AI agent better! :sweat_smile: 
```suggestion
        if (!Feature::isActive('v6.8.0.0')) {
```

shyim committed:
Update src/Core/Framework/Adapter/AdapterException.php

Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

mitelg commented:
thanks for your contribution @shyim 👍🎉💙
